core/src/main/java/org/elasticsearch/common/network/NetworkService.java

```diff
@@ -95,99 +92,86 @@ public void addCustomNameResolver(CustomNameResolver customNameResolver) {
     }
 
 
-    public InetAddress resolveBindHostAddress(String bindHost) throws IOException {
-        return resolveBindHostAddress(bindHost, InetAddress.getLoopbackAddress().getHostAddress());
+    public InetAddress[] resolveBindHostAddress(String bindHost) throws IOException {
+        return resolveBindHostAddress(bindHost, "_local_");
     }
 
-    public InetAddress resolveBindHostAddress(String bindHost, String defaultValue2) throws IOException {
-        return resolveInetAddress(bindHost, settings.get(GLOBAL_NETWORK_BINDHOST_SETTING, settings.get(GLOBAL_NETWORK_HOST_SETTING)), defaultValue2);
+    private InetAddress[] resolveBindHostAddress(String bindHost, String defaultValue2) throws IOException {
+        /** TODO: move this leniency out */
+        if (bindHost == null) {
+            bindHost = settings.get(GLOBAL_NETWORK_BINDHOST_SETTING, settings.get(GLOBAL_NETWORK_HOST_SETTING));
+        }
+        if (bindHost == null) {
+            bindHost = defaultValue2;
+        }
+        return resolveInetAddress(bindHost);
     }
 
+    // TODO: needs to be InetAddress[]
     public InetAddress resolvePublishHostAddress(String publishHost) throws IOException {
-        InetAddress address = resolvePublishHostAddress(publishHost,
-                InetAddress.getLoopbackAddress().getHostAddress());
-        // verify that its not a local address
-        if (address == null || address.isAnyLocalAddress()) {
-            address = NetworkUtils.getFirstNonLoopbackAddress(NetworkUtils.StackType.IPv4);
-            if (address == null) {
-                address = NetworkUtils.getFirstNonLoopbackAddress(NetworkUtils.getIpStackType());
-                if (address == null) {
-                    address = NetworkUtils.getLocalAddress();
-                    if (address == null) {
-                        return NetworkUtils.getLocalhost(NetworkUtils.StackType.IPv4);
-                    }
-                }
-            }
-        }
+        InetAddress address = resolvePublishHostAddress(publishHost, "_local_");
```

Comment:
I was concurrently cleaning up this code, because it was still too much for me. Please see the latest commit.